nvflare/edge/assessors/buff_device_manager.py

```diff
@@ -95,11 +220,16 @@ def remove_devices_from_used(self, devices: Set[str], fl_ctx) -> None:
         for device_id in devices:
             self.used_devices.pop(device_id, None)
 
-    def has_enough_devices(self, fl_ctx) -> bool:
+    def has_enough_devices_and_clients(self, fl_ctx) -> bool:
         num_holes = self.device_selection_size - len(self.current_selection)
         usable_devices = set(self.available_devices.keys()) - set(self.used_devices.keys())
         num_usable_devices = len(usable_devices)
-        return num_usable_devices >= num_holes
+        if num_usable_devices < num_holes:
+            return False
+
+        # Further check if we have enough clients
+        unique_clients = set(self.device_client_map.values())
+        return len(unique_clients) >= self.initial_min_client_num
```

Comment:
why do we need "initial_min_client_num"?

Reply:
we need to let server know how many client it needs to wait for - otherwise let's say we need 100 device from 4 clients, without knowing how many client are out there, server will just assign all 100 devices to the first client reporting to it - then there won't be "balanced" selection

Reply:
got it, we are checking this every round, so this is not just "initial" it's "min_client_num"?

Reply:
right, but it matters more at initial phase since at that time leaf nodes reports one after another, while later phase they only update the available device list - not quite likely that we lose a leaf node entirely